# Model Evoluation

# Model Evoluation - bu o'rgatilgan model qanchalik yaxshi ishlayotganini o'lchash jarayoni 

# Evaluation - bu metriklar yordamida baholash 

# Bir nechta muammolarni oldini olish maqsadida cross validation baholash mezonini tadbiq qilamiz

# CROSS VALIDATION 

# Cross Validation - bu modelni bitta train test bo'lishiga qarab qo'ymasdan uni bir nechta marta turli bo'linmalarda sinab ko'rish jarayoni 

# Nega Cross Validation kerak?
- agarda datasetni faqat bir marta train test qilib baxolasak, u holda natija tasodifga bog'liq bo'lishi mumkin, model overfitting yoki underfitting qilganini bilib olishimiz qiyin

Cross Validation esa:
- barqaror baho beradi
- kichik datasetlarda juda foydali
- modelni tanlashda adolatli solishtirish imkonini beradi

# Ishlashi

- Dataset bir nechqa qism(fold)larga bbolinadi

-Har safar bitta qism test qilinadi 

-qolganlari train qilinadi


-natijada o'rtacha baxolash qiymati olinadi


# Cross Validation turlari 

- K-Fold
- Stratified K-Fold
- Leave-one-out(LOO)
- Time Series Split

In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

df=pd.read_csv(r"C:\Users\Jahongir\Desktop\Project_One\Data\Raw_Data\titanic.csv")

In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
df = df[["Survived", "Pclass", "Sex", "Age", "Fare"]]

df["Age"].fillna(df["Age"].median(),inplace=True)
df["Sex"] = df["Sex"].map({"male": 0, "female": 1})

C:\Users\Jahongir\AppData\Local\Temp\ipykernel_12408\3894304004.py:3: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df["Age"].fillna(df["Age"].median(),inplace=True)


In [5]:
x = df.drop("Survived", axis=1)
y = df["Survived"]

# KFOLD

In [6]:
kfold = KFold(
    n_splits=5,  # K=5 Tavsiya qilinadi 5dan 10 tagacha
    shuffle=True,  # datasetni aralashtirish
    random_state=42
)

In [7]:
pipeline = Pipeline(steps=[
    ("scaler", StandardScaler()),
    ("model", LogisticRegression(max_iter=1000))
])

In [10]:
x["Age"] = x["Age"].fillna(x["Age"].median())

In [11]:
x.isnull().sum()

Pclass    0
Sex       0
Age       0
Fare      0
dtype: int64

# CROSS VALIDATION yordamida baaaholash

In [12]:
scores = cross_val_score(
    estimator=pipeline,
    X=x,
    y=y,
    cv=kfold,
    scoring="accuracy"
) 

In [13]:
print("Har bir fold accuracy natijasi:", scores)
print("O'rtacha accuracy:", scores.mean())
print("Standart og'ish:", scores.std())

Har bir fold accuracy natijasi: [0.79329609 0.7752809  0.81460674 0.74719101 0.78651685]
O'rtacha accuracy: 0.7833783190006905
Standart og'ish: 0.02217438343318582


# Stratified K-Fold Validation

In [15]:
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

scores = cross_val_score(
    pipeline,
    x,
    y,
    cv=skf,
    scoring= "accuracy"
)
scores.mean()

np.float64(0.7878664239532986)

# Stratified kfol va  Kfold bilan farqi nimada 

- Mummo Klasslar imblanance bo`lishi 
- Misol uchun : Survived ustunida 0 lar miqdori 62% ga teng 1 lar esa 38 % ga teng 

Agarda oddiy k fold bilan buni tekshirsak 
- bazi foldlarda 1 juda kam yoki umuman bo`lmasligi mumkin 
- model notogri baholanadi

# Shu muammoni oldini olgan holda stratified k fold klasslar nisbatini saqlab qoladi 

# Xulosa : 
- Balancli Datasetlar uchun - K-Fold 
- Imbalance Dataset bo`lsa - Stratified F-Fold 
- Classification model uchun - Stratified F-Fold 
- Regression model uchun - K-Fold 